src/webserver/http.rs

```diff
@@ -432,6 +432,53 @@ async fn serve_file(
         })
 }
 
+/// Fallback handler for when a file could not be served
+///
+/// Recursively traverses upwards in the request's path, looking for a `404.sql` to call as the
+// fallback handler. Fails if none is found, or if there is an issue while running the `404.sql`.
+async fn serve_fallback(
+    service_request: ServiceRequest,
+    original_err: actix_web::Error,
+) -> actix_web::Result<ServiceResponse> {
+    let catch_all = "404.sql";
+
+    let req_path = req_path(&service_request);
+    let mut maybe_fallback_path = req_path.clone().into_owned();
+    log::debug!("Trying to find a {catch_all:?} handler for {maybe_fallback_path:?}");
+
+    // Find the indeces of each char which follows a directroy separator (`/`). Also consider the 0
+    // index, as we also have to try to check the empty path, for a root dir `404.sql`.
+    for idx in req_path
+        .rmatch_indices('/')
+        .map(|(idx, _)| idx + 1)
+        .chain(std::iter::once(0))
+    {
+        // Remove the trailing substring behind the current `/`, and append `404.sql`.
+        maybe_fallback_path.truncate(idx);
+        maybe_fallback_path.push_str(&catch_all);
+
+        // Check if `maybe_fallback_path` actually exists, if not, skip to the next round (which
+        // will check `maybe_fallback_path`s parent directory for fallback handler).
+        let sql_path = PathBuf::from(&maybe_fallback_path);
+        let app_state: &web::Data<AppState> = service_request.app_data().expect("app_state");
```

Comment:
we should do that just once, out of the loop

Reply:
Done